bokehjs/src/lib/models/glyphs/image.ts

```diff
@@ -48,6 +52,55 @@ export class ImageView extends XYGlyphView {
     }
   }
 
+  _index_data(): SpatialIndex {
+    const points = [];
+    for (let i = 0, end = this._x.length; i < end; i++) {
+      const [l, r, t, b] = this._lrtb(i);
+      if (isNaN(l+r+t+b) || !isFinite(l+r+t+b)) {
+        continue;
+      }
+      points.push({minX: l, minY: b, maxX: r, maxY: t, i});
+    }
+    return new SpatialIndex(points);
+  }
+
+  _lrtb(i: number) : [number, number, number, number]{
+    const l = this._x[i]
+    const r = l + this._dw[i]
+    const b = this._y[i]
+    const t = b + this._dh[i]
+    return [l, r, t, b]
+  }
+
+  _image_index(index : number, x: number, y : number) : [number, number, number, number] {
+    const [l,r,t,b] = this._lrtb(index);
+    const width = this._width[index]
+    const height = this._height[index]
+    const dx = (r - l) / width;
+    const dy = (t - b) / height;
+    const dim1 = Math.floor((x-l) / dx)
+    const dim2 = Math.floor((y-b) / dy)
+    return [index, dim1,  dim2, (dim2 * width) + dim1]
+  }
```

Comment:
This should get a comment about the 4-tuple return at least. Alternatively would it be feasible to return a structure?

Reply:
Something like:

 ```javascript
{'index':index, 'dim1':dim1, 'dim2':dim2, 'flat_index':dim2 * width) + dim1}
```?

Sure, I think that makes sense and will work and that having explicit keys isn't a bad idea. I guess it means the property definition will need to be updated accordingly.

Reply:
Now using a structure in 4e9e2041b